activerecord/lib/active_record/connection_adapters/abstract/transaction.rb

```diff
@@ -44,21 +45,23 @@ def add_record(record); end
 
     class Transaction #:nodoc:
 
-      attr_reader :connection, :state, :records, :savepoint_name
+      attr_reader :connection, :state, :savepoint_name, :transactional_records
       attr_writer :joinable
 
       def initialize(connection, options)
         @connection = connection
         @state = TransactionState.new
-        @records = []
+        @transactional_records, @records = [], []
```

Comment:
Maybe better to use `Set` instead of `Array` to track uniqueness automatically. It may perform faster.